core/finality-grandpa/src/lib.rs

```diff
@@ -275,10 +275,15 @@ fn commit_topic<B: BlockT>(set_id: u64) -> B::Hash {
 	<<B::Header as HeaderT>::Hashing as HashT>::hash(format!("{}-COMMITS", set_id).as_bytes())
 }
 
-impl<B: BlockT, S: network::specialization::NetworkSpecialization<B>, H: ExHashT> Network<B> for NetworkBridge<B, S, H> {
+impl<B: BlockT, S: network::specialization::NetworkSpecialization<B>,> Network<B> for NetworkBridge<B, S> {
 	type In = mpsc::UnboundedReceiver<ConsensusMessage>;
 	fn messages_for(&self, round: u64, set_id: u64) -> Self::In {
-		self.service.consensus_gossip().write().messages_for(message_topic::<B>(round, set_id))
+		let (tx, rx) = sync::oneshot::channel();
+		self.service.with_gossip(move |gossip, _| {
+			let inner_rx = gossip.messages_for(message_topic::<B>(round, set_id));
+			let _ = tx.send(inner_rx);
+		});
+		rx.wait().ok().expect("1. Network is running, 2. it should handle the above closure successfully")
```

Comment:
I don't like waiting on the receiver here; would prefer it if the `type In` accounted for the receiver.
The assumption that the network is running is a bit strong as well and could easily be undercut.

Reply:
also worth noting that in `futures 0.3` (which we will eventually port to when it's more stable) has removed `wait` entirely.

Reply:
I agree with you, I don't like the `wait` either. On the plus side, it's better than the previous `write` on the `RwLock`, since that equally would block the current thread, and the channel now gives us a clear hook to better integrate the workflow into the surrounding futures context, as you suggest. Is this perhaps something that could be addressed in a follow-up?

With regards to the "correctness" of `wait`, my [understanding](https://docs.rs/futures/0.1.25/futures/future/trait.Future.html#method.wait) is that the method blocks the current thread, not just the current task, so it would appear "correct", with the drawback off-course that it blocks all other tasks running on the same thread in the pool. The call to `write` that this replaces had a similar effect...